documentation/interrupt_handling.md

```diff
@@ -1,23 +1,79 @@
-  configuration of hypervisor
-* HVBAR or VBAR (secure and non-secure) for ARMv7, depending on
-  configuration of hypervisor
-
-The processor is configured to use:
-* Monitor vector for FIQ exceptions while SCR_NS is set and state vector
-  when SCR_NS is cleared
-* Monitor vector for SMC exceptions
-* State vector for IRQ exceptions
+Exceptions, Interrupts and World Contextes
+==========================================
+
+This document describes describes how optee_os handles switches of
+world execution context based on SMC exceptions and interrupt notifications
+Interrupt notifications are IRQ/FIQ exceptions which may also imply switching
+of world execution context: non secure to secure, or secure to non secure.
+
+## Content
+
+1. [Context Switches and Exceptions](#1-context-switches-and-exceptions)
+2. [Normal World Invokes the optee_os Through a SMC](#2-normal-world-invokes-the-optee_os-through-a-smc)
+3. [Deliver a Non Secure Interrupt to the Non Secure World](#3-deliver-a-non-secure-interrupt-to-the-non-secure-world)
+4. [Deliver a Secure Interrupt to the Secure World](#4-deliver-a-secure-interrupt-to-the-secure-world)
+5. [TEE Thread Scheduling](#5-tee-thread-scheduling)
+
+# 1. Context Switches and Exceptions
+
+## 1.1. Usecases of world context switch
+
+When the normal world invokes the secure world, the normal world executes
+a SMC instruction. The SMC exception is always trapped by the Monitor. If
+the related service targets the trusted OS, the monitor invokes optee_os.
+When the secure world invokes the normal world, optee_os executes a SMC
+caught from the Monitor that will invoke the normal world.
+
+When a secure interrupt is signaled by the ARM GIC, it shall reach the
+optee_os interrupt exception vector. If the secure world was executing,
+optee_os handles straight the interrupt from it exception vector. If the
+normal world was executing when the secure interrupt raised, the Monitor
+vector must handles the exception and invoke the optee_os to serve the
+interrupt.
+
+When a non secure interrupt is signaled by the ARM GIC, it shall reach the
+normal world interrupt exception vector. If the non secure world was
+executing, it handles straight the exception from its exception vector. If
+the secure world was executing when the non secure interrupt raised,
+optee_os must invoke the Monitor to switch back to normal world so the
+interrupt can be served in normal world.
+
+## 1.2. Core Exception Vectors
+
+Monitor vector is VBAR_EL3 in AArch64 and MVBAR in ARMv7/AArch32. Monitor
+can be reached while non secure or secure world is executing. The executing
+secure state is known to the Monitor through the SCR_NS.
+
+Monitor can be reached from a SMC exception, an IRQ or FIQ exception
+(so-called interrupts) and from asynchronous aborts. Obviously monitor
+aborts (data, prefetch, undef) are local to the Monitor execution.
+Monitor can be internal (core/arch/arm/sm) to optee_os or external
```

Comment:
IMHO, it would be good to notice that OP-TEE's monitor `core/arch/arm/sm` is used on ARMv7, and arm-trusted-firmware on ARMv8 (people will definitely ask why there are two of them and what relationship are between them)